File: central-discovery-server/internal/discovery/acme.go
package discovery

import (
	"bytes"
	"context"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

const (
	acmeJWSContentType             = "application/jose+json"
	pemCertificateChainContentType = "application/pem-certificate-chain"
)

type http01ChallengeServer struct {
	address string
	tokens  map[string]string
	server  *http.Server
}

func newHTTP01ChallengeServer(address string) *http01ChallengeServer {
	return &http01ChallengeServer{
		address: address,
		tokens:  map[string]string{},
	}
}

func (s *http01ChallengeServer) setToken(token string, keyAuthorization string) {
	s.tokens[token] = keyAuthorization
}

func (s *http01ChallengeServer) start() error {
	mux := http.NewServeMux()
	mux.HandleFunc("/.well-known/acme-challenge/", func(w http.ResponseWriter, r *http.Request) {
		token := strings.TrimPrefix(r.URL.Path, "/.well-known/acme-challenge/")
		value, ok := s.tokens[token]
		if !ok {
			http.NotFound(w, r)
			return
		}
		w.Header().Set("Content-Type", "text/plain")
		_, _ = w.Write([]byte(value))
	})
	s.server = &http.Server{
		Addr:              s.address,
		Handler:           mux,
		ReadHeaderTimeout: 10 * time.Second,
	}
	listener, err := net.Listen("tcp", s.address)
	if err != nil {
		return err
	}
	go func() {
		if err := s.server.Serve(listener); err != nil && err != http.ErrServerClosed {
			logError("ACME HTTP-01 challenge server stopped: %v", err)
		}
	}()
	return nil
}

func (s *http01ChallengeServer) stop(ctx context.Context) error {
	if s.server == nil {
		return nil
	}
	return s.server.Shutdown(ctx)
}

type nativeACMEClient struct {
	directoryURL    string
	accountKeyPath  string
	certKeyPath     string
	certificatePath string
	email           string
	profile         string
	httpClient      *http.Client
	accountKey      *rsa.PrivateKey
	certKey         *rsa.PrivateKey
	directory       acmeDirectory
	kid             string
}

type acmeDirectory struct {
	NewNonce   string                     `json:"newNonce"`
	NewAccount string                     `json:"newAccount"`
	NewOrder   string                     `json:"newOrder"`
	Meta       map[string]json.RawMessage `json:"meta"`
}

func newNativeACMEClient(config Config) *nativeACMEClient {
	return &nativeACMEClient{
		directoryURL:    config.ACMEDirectoryURL,
		accountKeyPath:  filepath.Join(config.CertDir, "acme-account.key"),
		certKeyPath:     filepath.Join(config.CertDir, "central-discovery-ip.key"),
		certificatePath: filepath.Join(config.CertDir, "central-discovery-ip.crt"),
		email:           config.ACMEAccountEmail,
		profile:         config.ACMEProfile,
		httpClient:      &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *nativeACMEClient) issueIPCertificate(ctx context.Context, ipAddress string, challengeServer *http01ChallengeServer) error {
	var err error
	c.accountKey, err = loadOrCreateRSAKey(c.accountKeyPath)
	if err != nil {
		return err
	}
	c.certKey, err = loadOrCreateRSAKey(c.certKeyPath)
	if err != nil {
		return err
	}
	if err := c.getJSON(ctx, c.directoryURL, &c.directory); err != nil {
		return err
	}
	if err := c.ensureProfileAvailable(); err != nil {
		return err
	}
	if err := c.newAccount(ctx); err != nil {
		return err
	}
	order, err := c.newOrder(ctx, ipAddress)
	if err != nil {
		return err
	}
	if len(order.Authorizations) == 0 {
		return fmt.Errorf("ACME order did not include authorizations")
	}
	if err := c.completeAuthorization(ctx, order.Authorizations[0], challengeServer); err != nil {
		return err
	}
	csr, err := buildIPCSR(c.certKey, ipAddress)
	if err != nil {
		return err
	}
	finalized, err := c.post(ctx, order.Finalize, map[string]string{"csr": b64url(csr)})
	if err != nil {
		return err
	}
	if finalized.OrderURL != "" {
		order.OrderURL = finalized.OrderURL
	}
	order, err = c.pollOrder(ctx, order.OrderURL)
	if err != nil {
		return err
	}
	pemChain, err := c.postAsGetText(ctx, order.Certificate)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(c.certificatePath), 0o700); err != nil {
		return err
	}
	return os.WriteFile(c.certificatePath, []byte(pemChain), 0o644)
}

func (c *nativeACMEClient) getJSON(ctx context.Context, url string, target interface{}) error {
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	response, err := c.httpClient.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	if response.StatusCode >= 400 {
		body, _ := io.ReadAll(response.Body)
		return fmt.Errorf("ACME GET failed %d: %s", response.StatusCode, strings.TrimSpace(string(body)))
	}
	return json.NewDecoder(response.Body).Decode(target)
}

func (c *nativeACMEClient) newAccount(ctx context.Context) error {
	payload := map[string]interface{}{"termsOfServiceAgreed": true}
	if c.email != "" {
		payload["contact"] = []string{"mailto:" + c.email}
	}
	response, err := c.postJWS(ctx, c.directory.NewAccount, payload, true, "")
	if err != nil {
		return err
	}
	if response.Location == "" {
		return fmt.Errorf("ACME account response did not include Location")
	}
	c.kid = response.Location
	return nil
}

type acmeOrder struct {
	OrderURL       string   `json:"-"`
	Status         string   `json:"status"`
	Authorizations []string `json:"authorizations"`
	Finalize       string   `json:"finalize"`
	Certificate    string   `json:"certificate"`
}

func (c *nativeACMEClient) newOrder(ctx context.Context, ipAddress string) (acmeOrder, error) {
	payload := map[string]interface{}{
		"identifiers": []map[string]string{{"type": "ip", "value": ipAddress}},
	}
	if c.profile != "" {
		payload["profile"] = c.profile
	}
	response, err := c.post(ctx, c.directory.NewOrder, payload)
	if err != nil {
		return acmeOrder{}, err
	}
	if response.Order.OrderURL == "" {
		return acmeOrder{}, fmt.Errorf("ACME order response did not include Location")
	}
	return response.Order, nil
}

type acmeAuthorization struct {
	Status     string          `json:"status"`
	Challenges []acmeChallenge `json:"challenges"`
}

type acmeChallenge struct {
	Type   string `json:"type"`
	URL    string `json:"url"`
	Token  string `json:"token"`
	Status string `json:"status"`
}

func (c *nativeACMEClient) completeAuthorization(ctx context.Context, authorizationURL string, challengeServer *http01ChallengeServer) error {
	var authz acmeAuthorization
	if err := c.postAsGet(ctx, authorizationURL, &authz); err != nil {
		return err
	}
	var challenge *acmeChallenge
	for i := range authz.Challenges {
		if authz.Challenges[i].Type == "http-01" {
			challenge = &authz.Challenges[i]
			break
		}
	}
	if challenge == nil {
		return fmt.Errorf("ACME server did not offer HTTP-01 for IP certificate")
	}
	thumbprint, err := c.jwkThumbprint()
	if err != nil {
		return err
	}
	challengeServer.setToken(challenge.Token, challenge.Token+"."+thumbprint)
	if _, err := c.post(ctx, challenge.URL, map[string]interface{}{}); err != nil {
		return err
	}
	deadline := time.Now().Add(90 * time.Second)
	for time.Now().Before(deadline) {
		if err := waitForPollInterval(ctx); err != nil {
			return err
		}
		if err := c.postAsGet(ctx, authorizationURL, &authz); err != nil {
			return err
		}
		if authz.Status == "valid" {
			return nil
		}
		if authz.Status == "invalid" {
			raw, _ := json.Marshal(authz)
			return fmt.Errorf("ACME authorization failed: %s", raw)
		}
	}
	return fmt.Errorf("timed out waiting for ACME authorization")
}

func (c *nativeACMEClient) pollOrder(ctx context.Context, orderURL string) (acmeOrder, error) {
	deadline := time.Now().Add(90 * time.Second)
	for time.Now().Before(deadline) {
		var order acmeOrder
		if err := c.postAsGet(ctx, orderURL, &order); err != nil {
			return acmeOrder{}, err
		}
		order.OrderURL = orderURL
		if order.Status == "valid" && order.Certificate != "" {
			return order, nil
		}
		if order.Status == "invalid" {
			raw, _ := json.Marshal(order)
			return acmeOrder{}, fmt.Errorf("ACME order failed: %s", raw)
		}
		if err := waitForPollInterval(ctx); err != nil {
			return acmeOrder{}, err
		}
	}
	return acmeOrder{}, fmt.Errorf("timed out waiting for ACME certificate")
}

type acmePostResponse struct {
	Location string
	OrderURL string
	Order    acmeOrder
	Body     []byte
}

func (c *nativeACMEClient) post(ctx context.Context, url string, payload interface{}) (acmePostResponse, error) {
	response, err := c.postJWS(ctx, url, payload, false, "")
	if err != nil {
		return acmePostResponse{}, err
	}
	if len(response.Body) > 0 {
		if err := json.Unmarshal(response.Body, &response.Order); err != nil {
			return acmePostResponse{}, err
		}
	}
	response.Order.OrderURL = response.Location
	response.OrderURL = response.Location
	return response, nil
}

func (c *nativeACMEClient) postAsGet(ctx context.Context, url string, target interface{}) error {
	response, err := c.postJWS(ctx, url, nil, false, "")
	if err != nil {
		return err
	}
	if len(response.Body) == 0 {
		return nil
	}
	return json.Unmarshal(response.Body, target)
}

func (c *nativeACMEClient) postAsGetText(ctx context.Context, url string) (string, error) {
	response, err := c.postJWS(ctx, url, nil, false, pemCertificateChainContentType)
	if err != nil {
		return "", err
	}
	return string(response.Body), nil
}

func (c *nativeACMEClient) postJWS(ctx context.Context, url string, payload interface{}, useJWK bool, accept string) (acmePostResponse, error) {
	body, err := c.jws(ctx, url, payload, useJWK)
	if err != nil {
		return acmePostResponse{}, err
	}
	encoded, err := json.Marshal(body)
	if err != nil {
		return acmePostResponse{}, err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(encoded))
	if err != nil {
		return acmePostResponse{}, err
	}
	request.Header.Set("Content-Type", acmeJWSContentType)
	if accept != "" {
		request.Header.Set("Accept", accept)
	}
	response, err := c.httpClient.Do(request)
	if err != nil {
		return acmePostResponse{}, err
	}
	defer response.Body.Close()
	raw, err := io.ReadAll(response.Body)
	if err != nil {
		return acmePostResponse{}, err
	}
	if response.StatusCode >= 400 {
		return acmePostResponse{}, fmt.Errorf("ACME request failed %d: %s", response.StatusCode, strings.TrimSpace(string(raw)))
	}
	return acmePostResponse{
		Location: response.Header.Get("Location"),
		Body:     raw,
	}, nil
}

func (c *nativeACMEClient) jws(ctx context.Context, url string, payload interface{}, useJWK bool) (map[string]string, error) {
	if c.accountKey == nil {
		return nil, fmt.Errorf("ACME account key is not initialized")
	}
	protected := map[string]interface{}{
		"alg": "RS256",
		"url": url,
	}
	nonce, err := c.nonce(ctx)
	if err != nil {
		return nil, err
	}
	protected["nonce"] = nonce
	if useJWK {
		protected["jwk"] = c.jwk()
	} else {
		if c.kid == "" {
			return nil, fmt.Errorf("ACME account is not initialized")
		}
		protected["kid"] = c.kid
	}
	protectedJSON, err := json.Marshal(protected)
	if err != nil {
		return nil, err
	}
	protected64 := b64url(protectedJSON)
	payload64 := ""
	if payload != nil {
		payloadJSON, err := json.Marshal(payload)
		if err != nil {
			return nil, err
		}
		payload64 = b64url(payloadJSON)
	}
	signingInput := []byte(protected64 + "." + payload64)
	digest := sha256.Sum256(signingInput)
	signature, err := rsa.SignPKCS1v15(rand.Reader, c.accountKey, crypto.SHA256, digest[:])
	if err != nil {
		return nil, err
	}
	return map[string]string{
		"protected": protected64,
		"payload":   payload64,
		"signature": b64url(signature),
	}, nil
}

func (c *nativeACMEClient) nonce(ctx context.Context) (string, error) {
	request, err := http.NewRequestWithContext(ctx, http.MethodHead, c.directory.NewNonce, nil)
	if err != nil {
		return "", err
	}
	response, err := c.httpClient.Do(request)
	if err != nil {
		return "", err
	}
	defer response.Body.Close()
	nonce := response.Header.Get("Replay-Nonce")
	if nonce == "" {
		return "", fmt.Errorf("ACME server did not return a nonce")
	}
	return nonce, nil
}

func (c *nativeACMEClient) jwk() map[string]string {
	numbers := c.accountKey.PublicKey
	return map[string]string{
		"kty": "RSA",
		"n":   b64url(numbers.N.Bytes()),
		"e":   b64url(bigIntBytes(numbers.E)),
	}
}

func (c *nativeACMEClient) jwkThumbprint() (string, error) {
	jwk := c.jwk()
	canonical, err := json.Marshal(map[string]string{
		"e":   jwk["e"],
		"kty": jwk["kty"],
		"n":   jwk["n"],
	})
	if err != nil {
		return "", err
	}
	digest := sha256.Sum256(canonical)
	return b64url(digest[:]), nil
}

func (c *nativeACMEClient) ensureProfileAvailable() error {
	if c.profile == "" {
		return nil
	}
	raw, ok := c.directory.Meta["profiles"]
	if !ok {
		return nil
	}
	var profiles map[string]interface{}
	if err := json.Unmarshal(raw, &profiles); err != nil {
		return nil
	}
	if _, ok := profiles[c.profile]; ok {
		return nil
	}
	available := make([]string, 0, len(profiles))
	for profile := range profiles {
		available = append(available, profile)
	}
	sort.Strings(available)
	return fmt.Errorf("configured ACME profile %q is not advertised by the ACME directory; available profiles: %s", c.profile, strings.Join(available, ", "))
}

func b64url(raw []byte) string {
	return base64.RawURLEncoding.EncodeToString(raw)
}

func waitForPollInterval(ctx context.Context) error {
	timer := time.NewTimer(2 * time.Second)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

func bigIntBytes(value int) []byte {
	if value == 0 {
		return []byte{0}
	}
	var result []byte
	for value > 0 {
		result = append([]byte{byte(value & 0xff)}, result...)
		value >>= 8
	}
	return result
}

func logError(format string, args ...interface{}) {
	log.Printf(format, args...)
}

File: provider-server/provider/network_setup/acme.py
import asyncio
import base64
import hashlib
import json
import time
from pathlib import Path
from typing import Any

import aiohttp
from aiohttp import web
from cryptography.hazmat.primitives.asymmetric import padding, rsa

from .certs import build_ip_csr, load_or_create_rsa_key

ACME_JWS_CONTENT_TYPE = "application/jose+json"
PEM_CERTIFICATE_CHAIN_CONTENT_TYPE = "application/pem-certificate-chain"


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


class AcmeRequestError(RuntimeError):
    """Raised when the ACME server rejects a protocol or account request."""


class Http01ChallengeServer:
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port
        self._tokens: dict[str, str] = {}
        self._runner: web.AppRunner | None = None

    def set_token(self, token: str, key_authorization: str) -> None:
        self._tokens[token] = key_authorization

    async def start(self) -> None:
        app = web.Application()
        app.router.add_get(
            "/.well-known/acme-challenge/{token}", self._handle_challenge
        )
        self._runner = web.AppRunner(app)
        await self._runner.setup()
        site = web.TCPSite(self._runner, self.host, self.port)
        await site.start()

    async def stop(self) -> None:
        if self._runner is not None:
            await self._runner.cleanup()
            self._runner = None

    async def _handle_challenge(self, request: web.Request) -> web.Response:
        token = request.match_info["token"]
        value = self._tokens.get(token)
        if value is None:
            return web.Response(status=404)
        return web.Response(text=value, content_type="text/plain")


class NativeAcmeClient:
    """Small ACME v2 client for Let's Encrypt IP HTTP-01 certificates."""

    def __init__(
        self,
        directory_url: str,
        account_key_path: Path,
        cert_key_path: Path,
        certificate_path: Path,
        email: str = "",
        profile: str = "shortlived",
        session: aiohttp.ClientSession | None = None,
    ):
        self.directory_url = directory_url
        self.account_key_path = account_key_path
        self.cert_key_path = cert_key_path
        self.certificate_path = certificate_path
        self.email = email
        self.profile = profile
        self.session = session
        self.account_key: rsa.RSAPrivateKey | None = None
        self.cert_key: rsa.RSAPrivateKey | None = None
        self.directory: dict[str, Any] = {}
        self.kid: str | None = None

    async def issue_ip_certificate(
        self,
        ip_address: str,
        challenge_server: Http01ChallengeServer,
    ) -> None:
        owns_session = self.session is None
        session = self.session or aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=30)
        )
        try:
            self.session = session
            self.account_key = load_or_create_rsa_key(self.account_key_path)
            self.cert_key = load_or_create_rsa_key(self.cert_key_path)
            self.directory = await self._get_json(self.directory_url)
            self._ensure_profile_available()
            await self._new_account()
            order = await self._new_order(ip_address)
            await self._complete_authorization(
                order["authorizations"][0], challenge_server
            )
            csr_der = build_ip_csr(self.cert_key, ip_address)
            finalized = await self._post(order["finalize"], {"csr": _b64url(csr_der)})
            order = await self._poll_order(
                finalized.get("order_url") or order["order_url"]
            )
            pem = await self._post_as_get_text(order["certificate"])
            self.certificate_path.parent.mkdir(parents=True, exist_ok=True)
            self.certificate_path.write_text(pem)
            self.certificate_path.chmod(0o644)
        finally:
            if owns_session:
                await session.close()
                self.session = None

    async def _get_json(self, url: str) -> dict[str, Any]:
        assert self.session is not None
        async with self.session.get(url) as response:
            response.raise_for_status()
            return await response.json()

    async def _new_account(self) -> None:
        payload: dict[str, Any] = {"termsOfServiceAgreed": True}
        if self.email:
            payload["contact"] = [f"mailto:{self.email}"]
        response = await self._post(
            self.directory["newAccount"],
            payload,
            use_jwk=True,
        )
        self.kid = response["kid"]

    async def _new_order(self, ip_address: str) -> dict[str, Any]:
        payload: dict[str, Any] = {"identifiers": [{"type": "ip", "value": ip_address}]}
        if self.profile:
            payload["profile"] = self.profile
        response = await self._post(self.directory["newOrder"], payload)
        response["order_url"] = response["url"]
        return response

    async def _complete_authorization(
        self,
        authorization_url: str,
        challenge_server: Http01ChallengeServer,
    ) -> None:
        authz = await self._post_as_get(authorization_url)
        challenge = next(
            (
                item
                for item in authz.get("challenges", [])
                if item.get("type") == "http-01"
            ),
            None,
        )
        if challenge is None:
            raise RuntimeError("ACME server did not offer HTTP-01 for IP certificate")
        token = challenge["token"]
        challenge_server.set_token(token, f"{token}.{self._jwk_thumbprint()}")
        await self._post(challenge["url"], {})
        deadline = time.monotonic() + 90
        while time.monotonic() < deadline:
            await asyncio.sleep(2)
            authz = await self._post_as_get(authorization_url)
            status = authz.get("status")
            if status == "valid":
                return
            if status == "invalid":
                raise RuntimeError(f"ACME authorization failed: {authz}")
        raise RuntimeError("Timed out waiting for ACME authorization")

    async def _poll_order(self, order_url: str) -> dict[str, Any]:
        deadline = time.monotonic() + 90
        while time.monotonic() < deadline:
            order = await self._post_as_get(order_url)
            if order.get("status") == "valid" and order.get("certificate"):
                return order
            if order.get("status") == "invalid":
                raise RuntimeError(f"ACME order failed: {order}")
            await asyncio.sleep(2)
        raise RuntimeError("Timed out waiting for ACME certificate")

    async def _post_as_get(self, url: str) -> dict[str, Any]:
        return await self._post(url, None)

    async def _post_as_get_text(self, url: str) -> str:
        _headers, text = await self._post_jws(
            url,
            None,
            accept=PEM_CERTIFICATE_CHAIN_CONTENT_TYPE,
        )
        return text

    async def _post(
        self,
        url: str,
        payload: dict[str, Any] | None,
        use_jwk: bool = False,
    ) -> dict[str, Any]:
        headers, text = await self._post_jws(url, payload, use_jwk=use_jwk)
        data = json.loads(text) if text else {}
        if headers.get("Location"):
            data["url"] = headers["Location"]
        if use_jwk and headers.get("Location"):
            data["kid"] = headers["Location"]
        return data

    async def _post_jws(
        self,
        url: str,
        payload: dict[str, Any] | None,
        use_jwk: bool = False,
        accept: str | None = None,
    ) -> tuple[dict[str, str], str]:
        assert self.session is not None
        body = await self._jws(url, payload, use_jwk=use_jwk)
        headers = {"Content-Type": ACME_JWS_CONTENT_TYPE}
        if accept:
            headers["Accept"] = accept
        async with self.session.post(
            url,
            data=json.dumps(body, separators=(",", ":")).encode("utf-8"),
            headers=headers,
        ) as response:
            text = await response.text()
            if response.status >= 400:
                raise AcmeRequestError(f"ACME request failed {response.status}: {text}")
            return dict(response.headers), text

    async def _jws(
        self,
        url: str,
        payload: dict[str, Any] | None,
        use_jwk: bool = False,
    ) -> dict[str, str]:
        assert self.account_key is not None
        protected: dict[str, Any] = {
            "alg": "RS256",
            "nonce": await self._nonce(),
            "url": url,
        }
        if use_jwk:
            protected["jwk"] = self._jwk()
        else:
            if not self.kid:
                raise RuntimeError("ACME account is not initialized")
            protected["kid"] = self.kid
        protected64 = _b64url(
            json.dumps(protected, separators=(",", ":")).encode("utf-8")
        )
        payload64 = (
            ""
            if payload is None
            else _b64url(json.dumps(payload, separators=(",", ":")).encode("utf-8"))
        )
        signing_input = f"{protected64}.{payload64}".encode("ascii")
        signature = self.account_key.sign(
            signing_input,
            padding.PKCS1v15(),
            self._hash_algorithm(),
        )
        return {
            "protected": protected64,
            "payload": payload64,
            "signature": _b64url(signature),
        }

    async def _nonce(self) -> str:
        assert self.session is not None
        async with self.session.head(self.directory["newNonce"]) as response:
            response.raise_for_status()
            nonce = response.headers.get("Replay-Nonce")
            if not nonce:
                raise RuntimeError("ACME server did not return a nonce")
            return nonce

    def _jwk(self) -> dict[str, str]:
        assert self.account_key is not None
        numbers = self.account_key.public_key().public_numbers()
        return {
            "kty": "RSA",
            "n": _b64url(numbers.n.to_bytes((numbers.n.bit_length() + 7) // 8, "big")),
            "e": _b64url(numbers.e.to_bytes((numbers.e.bit_length() + 7) // 8, "big")),
        }

    def _jwk_thumbprint(self) -> str:
        jwk = self._jwk()
        canonical = json.dumps(
            {"e": jwk["e"], "kty": jwk["kty"], "n": jwk["n"]},
            sort_keys=True,
            separators=(",", ":"),
        ).encode("utf-8")
        return _b64url(hashlib.sha256(canonical).digest())

    def _ensure_profile_available(self) -> None:
        if not self.profile:
            return
        profiles = self.directory.get("meta", {}).get("profiles")
        if not isinstance(profiles, dict):
            return
        if self.profile in profiles:
            return
        available = ", ".join(sorted(str(profile) for profile in profiles))
        raise AcmeRequestError(
            f"Configured ACME profile '{self.profile}' is not advertised by "
            f"the ACME directory. Available profiles: {available or 'none'}."
        )

    @staticmethod
    def _hash_algorithm():
        from cryptography.hazmat.primitives import hashes

        return hashes.SHA256()
